### PR TITLE
Issue 191

### LINKED ISSUE #191: Server side allows to create overlapping service commitments
**Describe the bug**
Our POST /service_commitment api endpoint allows users to create service commitments with overlapping times. This is undesirable, because we want to prevent users from singing up for service shifts that have overlapping times.

**To Reproduce**
There are two ways to create overlapping commitments.
Option 1 - send overlapping commitments in the same POST request
1. Create two service sifts for the same time slots but for different shelters
```
curl --location '127.0.0.1:5000/service_shift' \
--header 'Authorization: 1234567890-developer-token' \
--header 'Content-Type: application/json' \
--data '[
    {"shelter_id":1111, "shift_start": 10, "shift_end": 20}, 
    {"shelter_id":2222, "shift_start": 10, "shift_end": 20}
]'
```
2. Note the service shift ids that got returned. Suppose they are ID1 and ID2
3. Create two commitments for the same user, one for shift with ID1 and another for shift with ID2
```
curl --location '127.0.0.1:5000/service_commitment' \
--header 'Authorization: 1234567890-developer-token' \
--header 'Content-Type: application/json' \
--data '[{"service_shift_id": ID1}, {"service_shift_id": ID2}]'
```
5. Both commitments will be successfully created

Option 2 - create commitments one at a time
1. Create two service sifts for the same time slots but for different shelters
```
curl --location '127.0.0.1:5000/service_shift' \
--header 'Authorization: 1234567890-developer-token' \
--header 'Content-Type: application/json' \
--data '[
    {"shelter_id":1111, "shift_start": 10, "shift_end": 20}, 
    {"shelter_id":2222, "shift_start": 10, "shift_end": 20}
]'
```
2. Note the service shift ids that got returned. Suppose they are ID1 and ID2
3. Create a commitment for service shift with ID1
```
curl --location '127.0.0.1:5000/service_commitment' \
--header 'Authorization: 1234567890-developer-token' \
--header 'Content-Type: application/json' \
--data '[{"service_shift_id": ID1}]'
```
4. Create a commitment for service shift with ID2
```
curl --location '127.0.0.1:5000/service_commitment' \
--header 'Authorization: 1234567890-developer-token' \
--header 'Content-Type: application/json' \
--data '[{"service_shift_id": ID2}]'
```
5. Both commitments will be successfully created

**Expected Behavior**
In both cases (Option 1 and Option 2), the second commitment should be rejected with a message 'Overlapping commitment'. The first commitment should be created.

**Suggested approach**
Issue #189, when completed, will ensure that service commitments will only be created for existing service_shift_id. To do this, the `add_service_commitments` function of use_cases/add_service_commitments.py will be modified: commitment_repo and shift_repo will be passed to the function as arguments. 

After issue #189 is resolved, we can improve the `add_service_commitments` function by checking the time slots associated with each commitment. Time slots are defined in the service shifts, and each commitment is associated with a specific shift id. Therefore, we can retrieve all service shifts (from the shift_repo) for the given list of commitments, check their time slots and see if any of them overlap. If they overlap, we reject the overlapping commitment that appears later in the list with a message 'overlapping commitments'.

Additionally, we need to make sure that the new commitments for this user do not overlap with any commitments that already exist in the database. Therefore, we also need to retrieve a list of all commitments the user has and check for overlap with them.

**Success criteria**

- [x] In the scenario described under Option 1 above, the first commitment should get created while the second one should not
- [x] In the scenario described under Option 2 above, the first commitment should get created while the second one should not.

### PR DESCRIPTION
**Pull request by strejoarciles**
Fixes #191 

I changed the service commitment functionality within the API endpoint for creating service commitments. Specifically, I modified the logic in the add_service_commitments function to prevent users from creating overlapping service commitments.

The change was necessary because the original implementation allowed users to sign up for service shifts with overlapping times, which is undesirable. The API was incorrectly accepting multiple commitments for the same time period, either when submitted in a single request or across multiple requests. This fix ensures that a user can only commit to non-overlapping service shifts.

I modified the add_service_commitments function in use_cases/add_service_commitments.py. The key changes include: retrieving the user's existing commitments from the database using fetch_service_commitments, checking for time overlaps between new commitments and existing ones, implementing a one-at-a-time processing approach, and adding robust error handling for test environments. The most important addition was the overlap detection logic that compares shift start and end times to prevent scheduling conflicts.

### CODE REVIEW

File: server/use_cases/add_service_commitments.py [hunk 1]
```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
-                shifts[i].shift_end > shifts[j].shift_start):
-                return True
+            if has_valid_time_attributes(
+                shifts[i]) and has_valid_time_attributes(
+                shifts[j]):
+                if (shifts[i].shift_start < shifts[j].shift_end and
+                    shifts[i].shift_end > shifts[j].shift_start):
+                    return True
     return False
+
```
Comment (Taktar) [inline server/use_cases/add_service_commitments.py:165]: Love the idea of checking validity of attributes to guard against mocks or malformed input. Nice touch! Maybe add unit tests for these new helpers to ensure edge cases (like partially missing fields or weird mocks) are covered.

File: server/use_cases/add_service_commitments.py [hunk 2]
```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
             }
         else:
             results[i] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Failed to insert commitment"
+            }
+    for i in valid_indexes:
+        if results[i] is None:
```
Comment (Taktar) [inline server/use_cases/add_service_commitments.py:137]: Nice catch updating the fallback None to an explicit response dict with an error message — this is definitely more informative for debugging and frontend feedback!

File: server/use_cases/add_service_commitments.py [hunk 3]
```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
             }
-
     return results
 
-
 def check_time_overlap(shifts):
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:145]: You have a check_time_overlap function, but this function does not get used. You should use it.

File: server/use_cases/add_service_commitments.py [hunk 4]
```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+        approved_shifts = [
+            valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+        for approved_shift in approved_shifts:
+            if has_valid_time_attributes(
+                current_shift) and has_valid_time_attributes(
+                approved_shift):
+                if (current_shift.shift_start < approved_shift.shift_end and
+                    current_shift.shift_end > approved_shift.shift_start):
+                    has_overlap = True
+                    break
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:102]: Instead of redefining the check for time overlap logic here, you should use the check_time_overlap function  you have in this file.

**Review by kate-holdener — CHANGES_REQUESTED** [review]
@strejoarciles I encourage you to test your code changes. When I send a POST request to create a service commitment for a valid shift ID, I get a 400 response status code with a message:
```
{
    "error": "Invalid shift ID: '134134' is not a valid ObjectId, it must be a 12-byte input or a 24-character hex string"
}
```
I added other inline comments for changes that need to be made.

File: server/use_cases/add_service_commitments.py [hunk 5]
```diff
@@ -39,52 +41,99 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+        shifts_to_check.extend(existing_user_shifts)
+        if check_time_overlap(shifts_to_check):
+            has_overlap = True
+        if not has_overlap and final_valid_indexes:
+            approved_shifts = [
+                valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+            shifts_to_check = [current_shift]
+            shifts_to_check.extend(approved_shifts)
+            if check_time_overlap(shifts_to_check):
+                has_overlap = True
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:84]:
```suggestion
```
It's redundant to do the check for overlap again. I understand why you are doing it - you want to make sure that it doesn't overlap with any of the commitments sent in the same request that you already approved. There is a simpler way to do this. See my suggestion later in this function.

File: server/use_cases/add_service_commitments.py [hunk 6]
```diff
@@ -39,52 +41,99 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+            final_valid_commitments.append(valid_commitments[i])
+            final_valid_indexes.append(idx)
+    if not final_valid_commitments:
+        for i in valid_indexes:
+            if results[i] is None:
+                results[i] = {
+                    "service_commitment_id": None,
+                    "success": False,
+                    "message": "Overlapping commitment"
+                }
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:101]:
```suggestion
```
I'm pretty sure that this is redundant. Since earlier, when you looped through valid_indexes, if the commitment was overlapping, you already inserted the response for the overlapping commitment into results[i] (on line 86)

File: server/use_cases/add_service_commitments.py [hunk 7]
```diff
@@ -39,52 +41,99 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
-                "success": False
+                "success": False,
+                "message": "Failed to insert commitment"
+            }
+    for i in valid_indexes:
+        if results[i] is None:
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:123]: This also seems redundant, you already marked the commitments as overlapping, when you discovered that they are overlapping (on line 86)
```suggestion
```

**Review by kate-holdener — CHANGES_REQUESTED** [review]
@strejoarciles 
Nice job on the progress. The code works for the cases specified in the issue.
There is a lot of redundant code that makes it difficult to read. I added my comments with proposed changes on how to simplify it.

Additionally, there is a bug: we are only considering overlaps if they are for the same shelter. We should not allow to create commitments for the same volunteer that overlap, regardless of whether they are in the same shelter or not (the same person cannot be volunteering in two different places at the same time).

File: server/use_cases/add_service_shifts.py [hunk 8]
```diff
@@ -39,11 +39,15 @@ def shift_add_use_case(repo, new_shifts):
...
         for other_shift in shifts[i + 1:]:
-            if (max(shift.shift_start, other_shift.shift_start) <
-                min(shift.shift_end, other_shift.shift_end)) and \
-               (shift.shelter_id == other_shift.shelter_id):
+            # Check time overlap
+            time_overlap = (
+                max(shift.shift_start, other_shift.shift_start) <
+                    min(shift.shift_end, other_shift.shift_end))
+            # Only consider it an overlap if at the same shelter
+            if time_overlap and shift.shelter_id == other_shift.shelter_id:
```
[the remark this answers is not in the document]
Comment (kate-holdener) [reply server/use_cases/add_service_shifts.py:51]: Oh, nevermind, I thought I was reviewing service_commitments code, but now I see that it's a change to add_service_shifts.py

File: server/use_cases/add_service_commitments.py [hunk 9]
```diff
@@ -39,52 +41,98 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        # Check for overlap with existing user shifts
+        for existing_shift in existing_user_shifts:
+            if (current_shift.shift_start < existing_shift.shift_end and
+                current_shift.shift_end > existing_shift.shift_start):
+                has_overlap = True
+                break
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:78]: What happened to calling check_time_overlap? You used to call it here, but now you duplicated the code that's in check_time_overlap in this function.

**Review by kate-holdener — CHANGES_REQUESTED** [review]
@strejoarciles You didn't address the code duplication problems I pointed out in my last review. Also, you added some more code duplication in this version, instead of reusing check_time_overlap function. Please address all the concerns I pointed out.

File: server/use_cases/add_service_commitments.py [hunk 10]
```diff
@@ -39,52 +41,98 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
...
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
             }
-
     return results
 
-
 def check_time_overlap(shifts):
```
Comment (kate-holdener) [inline server/use_cases/add_service_commitments.py:126]: You should change this function to be:
```suggestion
def check_time_overlap(shift_to_check, other_shifts):
```
Have this function return true if shift_to_check overlaps with any of the shifts in other_sifts.
Then use this function in your code
